### PR TITLE
Content data new 3

### PR DESCRIPTION
**Pull request by kolmanv**

### CODE REVIEW

File: lib/content_data/content_data.rb [hunk 1]
```diff
@@ -1,117 +1,28 @@
...
+  # aka decompress
+  def self.inflate(string)
+    zstream = Zlib::Inflate.new
+    buf = zstream.inflate(string)
+    zstream.finish
+    zstream.close
+    buf
   end
 
   # Unfortunately this class is used as mutable for now. So need to be carefull.
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:28]: Update the comment.

File: lib/content_data/content_data.rb [hunk 2]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
-        @contents = copy.contents.clone
-        @instances = copy.instances.clone
+        @db = copy.db  # db is cloned
       end
     end
 
-    def add_content(content)
-      @contents[content.checksum] = content
+    # getting a cloned data base
+    def db
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:43]: Don't use "db" notation it's confusing. Use more logical naming, i.e. contant mapping or something..

Comment (kolmanv) [inline lib/content_data/content_data.rb:43]: Rename this function to clone() or copy() what is accepted in Ruby.

File: lib/content_data/content_data.rb [hunk 3]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
-        Log.warning sprintf("Adding instance while it's" +
-                            " checksum %s does not exists.\n", instance.checksum)
-        Log.warning sprintf("%s\n", instance.to_s)
-        return false
-      elsif (@contents[instance.checksum].size != instance.size)
-        Log.warning 'File size different from content size while same checksum'
-        Log.warning instance.to_s
-        return false
-      end
+    def private_db
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:57]: Why is this needed? Delete this function.

File: lib/content_data/content_data.rb [hunk 4]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
     def content_exists(checksum)
-      @contents.key? checksum
+      @db.has_key?(checksum)
     end
 
-    # TODO(kolman): The semantics of thir merge is merge! change in all file.
-    def merge(content_data)
-      content_data.contents.values.each { |content|
-        add_content(content)
+    def instance_exists(full_path)
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:109]: Not efficient. Consider making a helper mapping.

File: lib/content_data/content_data.rb [hunk 5]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
+      @db.values.each { |content_db|
+        return true if content_db[1].has_key?(full_path)
       }
-      content_data.instances.values.each { |instance|
-        add_instance(instance)
+      false
+    end
+
+    def remove_instance(full_path)
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:116]: Same as above.
What is more important here, RAM or CPU time?

File: lib/content_data/content_data.rb [hunk 6]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
+      @db.keys.each { |checksum|
+        instances_db =  @db[checksum][1]
+        instances_db.delete(full_path)
+        @db.delete(checksum) if instances_db.empty?
       }
     end
 
     def ==(other)
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:124]: Are you sure we want to compare instances?
Maybe we want to compare only contents?

File: lib/content_data/content_data.rb [hunk 7]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
-        end
-      }
-      @instances.keys.each { |key|
-        if (@instances[key] != other.instances[key])
-          return false
-        end
-      }
-      return true
+      return false unless @db.size == other.private_db.size
+      @db.hash == other.private_db.hash
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:127]: What is ".hash" what it does?

File: lib/content_data/content_data.rb [hunk 8]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
     end
 
     def to_s
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:130]: Why do we need to_s?
Shouldn't we serialize the content data to binary representation?

File: lib/content_data/content_data.rb (hunk 1 above)
Comment (kolmanv) [inline lib/content_data/content_data.rb:28]: Please describe the data strcutrure here, i.e., 
hash => key ... , value ... types, and averything.

File: lib/content_data/content_data.rb [hunk 9]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
+      @db[checksum] = [size, Hash.new, time] unless @db.has_key?(checksum)
+    end
 
-      #override file if needed
-      @instances[key] = instance
+    def add_instance(checksum, size, server_name, device, full_path, modification_time)
+      full_path = "%s,%s,%s" % [server_name, device, full_path]
+      # remove instance from all contents except from the given checksum
+      # if no more instances then remove also the content data
+      @db.keys.each { |db_checksum|
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:73]: Not efficient. consider adding reverse map.

File: lib/content_data/content_data.rb [hunk 10]
```diff
@@ -1,117 +1,28 @@
...
-      "%s,%d,%s" % [@checksum, @size, ContentData.format_time(@first_appearance_time)]
-    end
-
-    def ==(other)
-      return (self.checksum.eql? other.checksum and
-              self.size.eql? other.size and
-              self.first_appearance_time.to_i.eql? other.first_appearance_time.to_i)
-    end
+  # aka compress
+  def self.deflate(string, level)
```
Comment (gshilin) [inline lib/content_data/content_data.rb:11]: Have you ever seen this?
http://www.ruby-doc.org/stdlib-1.9.3/libdoc/zlib/rdoc/Zlib/Deflate.html#method-c-deflate

Zlib::Deflate.deflate(string, level) is all you need, you don't have to copy standard function's content.

File: lib/content_data/content_data.rb [hunk 11]
```diff
@@ -1,117 +1,28 @@
...
-
-    def ==(other)
-      return (self.checksum.eql? other.checksum and
-              self.size.eql? other.size and
-              self.server_name.eql? other.server_name and
-              self.device.eql? other.device and
-              self.full_path.eql? other.full_path and
-              self.modification_time.to_i.eql? other.modification_time.to_i)
-    end
+  # aka decompress
```
Comment (gshilin) [inline lib/content_data/content_data.rb:19]: See above

File: lib/content_data/content_data.rb [hunk 12]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
-        @instances = copy.instances.clone
+        @db = copy.db  # db is cloned
       end
     end
 
-    def add_content(content)
-      @contents[content.checksum] = content
+    # getting a cloned data base
+    def db
+      clone_db = Hash.new
```
Comment (gshilin) [inline lib/content_data/content_data.rb:44]: All this stuff is called inject: http://ruby-doc.org/core-1.9.3/Enumerable.html#method-i-inject
@db.keys.inject({}) { |clone_db, checksum|
  ...
  clone_db
}

File: lib/content_data/content_data.rb [hunk 13]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
-        Log.warning 'File size different from content size while same checksum'
-        Log.warning instance.to_s
-        return false
-      end
+    def private_db
+      @db
+    end
 
-      key = instance.global_path
+    def set_db(db)
```
Comment (gshilin) [inline lib/content_data/content_data.rb:61]: Shouldn't this be called initialize?
If this is just setter, why don't use http://ruby-doc.org/core-1.9.3/Module.html#method-i-attr kind of declaration?
It makes your intention much clearer.

File: lib/content_data/content_data.rb [hunk 14]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
-      @contents.key? checksum
+      @db.has_key?(checksum)
     end
 
-    # TODO(kolman): The semantics of thir merge is merge! change in all file.
-    def merge(content_data)
-      content_data.contents.values.each { |content|
-        add_content(content)
+    def instance_exists(full_path)
+      @db.values.each { |content_db|
```
Comment (gshilin) [inline lib/content_data/content_data.rb:110]: This is called any?
http://ruby-doc.org/core-1.9.3/Enumerable.html#method-i-any-3F

File: lib/content_data/content_data.rb [hunk 15]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
+    def remove_instance(full_path)
+      @db.keys.each { |checksum|
+        instances_db =  @db[checksum][1]
+        instances_db.delete(full_path)
+        @db.delete(checksum) if instances_db.empty?
       }
     end
 
     def ==(other)
       return false if other == nil
```
Comment (gshilin) [inline lib/content_data/content_data.rb:125]: other.nil?

File: lib/content_data/content_data.rb [hunk 16]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
-        ret << instance.to_s << "\n"
+      return_str << "%d\n" % [keys.size]
+      return_str << contents_str
+      return_str << "%d\n" % [instances_counter]
+      return_str << instances_str
+      return_str
+    end
+
+    def to_a
+      returned_arr = []
```
Comment (gshilin) [inline lib/content_data/content_data.rb:158]: Again, this is Enumerable#inject.
And in #to_s too

File: lib/content_data/content_data.rb [hunk 17]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
+          returned_arr.push(instance[1][path])
+        }
+        returned_arr.push(instance[2])
       }
-      return ret
+      returned_arr
+    end
+
+    def from_a(arr)
+      if arr.nil?
```
Comment (gshilin) [inline lib/content_data/content_data.rb:175]: @db = nil and return if arr.nil?

File: lib/content_data/content_data.rb [hunk 18]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
+        @db = nil
+        return
+      end
+      @db = Hash.new
```
Comment (gshilin) [inline lib/content_data/content_data.rb:179]: Hash.new is the same as {}
The only difference that you _can_ supply default value different from nil.

File: lib/file_indexing/index_agent.rb [hunk 19]
```diff
@@ -69,107 +69,129 @@ def IndexAgent.get_correct_mtime(file)
...
-      # if there is a given DB then populate table with files
-      # that was already indexed on this server/device
-      if (otherDB != nil)
-        otherDB_contents.update(otherDB.contents)
-        otherDB.instances.each_value do |i|
-          next unless i.server_name == server_name #and i.device == @device
-          otherDB_table[i.full_path] = i
+      # TODO device support
+      def index(patterns, otherDB = nil)
+        abort "#{self.class}: DB not empty. Current implementation permits only one running of index" \
```
Comment (gshilin) [inline lib/file_indexing/index_agent.rb:74]: backslash is unnecessary here

File: lib/file_indexing/index_agent.rb [hunk 20]
```diff
@@ -69,107 +69,129 @@ def IndexAgent.get_correct_mtime(file)
...
+            unless @indexed_content.empty?
+
+        server_name = `hostname`.strip
+        permit_patterns = Array.new
```
Comment (gshilin) [inline lib/file_indexing/index_agent.rb:78]: Array.new is [] and Hash.new is {}

File: lib/file_indexing/index_agent.rb [hunk 21]
```diff
@@ -69,107 +69,129 @@ def IndexAgent.get_correct_mtime(file)
...
+        forbid_patterns = Array.new
+        otherDB_updated = Hash.new
+        #otherDB_table = Hash.new   # contains instances from given DB while full path name is a key and instance is a value
+        #otherDB_contents = Hash.new  # given DB contents
+
+        # if there is a given DB then populate table with files
+        # that was already indexed on this server/device
+        if (otherDB != nil)
```
Comment (gshilin) [inline lib/file_indexing/index_agent.rb:86]: if !otherDB.nil?

File: lib/file_indexing/index_agent.rb [hunk 22]
```diff
@@ -69,107 +69,129 @@ def IndexAgent.get_correct_mtime(file)
...
+        permit_patterns = patterns.positive_patterns
+        forbid_patterns = patterns.negative_patterns
 
-      # add files found by positive patterns
-      files = Array.new
-      permit_patterns.each_index do |i|
-        files = files | (collect(permit_patterns[i]));
-      end
+        # add files found by positive patterns
+        files = Array.new
```
Comment (gshilin) [inline lib/file_indexing/index_agent.rb:109]: inject

File: lib/file_utils/file_utils.rb [hunk 23]
```diff
@@ -198,67 +198,85 @@ def self.contet_data_command(command, cd_a, cd_b, dest_path)
...
+            content_time = instances[2]
+            not_found[checksum] = [size,
+                                  instances_db_cloned,
+                                  content_time]
+            warnings << "Warning: base content does not contains:'%s'" % checksum
+          end
+        }
+        not_found_cd.set_db(not_found)
+        Log.warning (warnings)
+        return not_found_cd
```
Comment (gshilin) [inline lib/file_utils/file_utils.rb:278]: return?

File: lib/content_data/content_data.rb [hunk 24]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
+        unless db_checksum.eql?(checksum)
+          instances_db =  @db[db_checksum][1]
+          instances_db.delete(full_path)
+          #@db.delete(db_checksum) if instances_db.empty?
+        end
+      }
+      instances = @db[checksum]
+
+      if instances.nil?
+        #TODO new appraoch - use this method without add_content?
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:83]: TODO is missing owner, i.e., should be TODO(kolman) so that I will know who to ask about the comment what did he though to do...

File: lib/content_data/content_data.rb [hunk 25]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
+        #Log.warning sprintf("Adding instance while it's" +
+        #                        " checksum %s does not exists.\n", checksum)
+        #Log.warning sprintf("instance full path:'%s'\n", full_path)
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:86]: Remove or un-comment those lines, make them debug_3 if you want to leave them as debug.

File: lib/content_data/content_data.rb [hunk 26]
```diff
@@ -119,91 +30,171 @@ def ==(other)
...
+          tmp_ind+=2
+          instance_db[instance_full_path]=instance_time
+        end
+        content_time = next_elem
+        @db[checksum] = [content_size,instance_db, content_time]
+        tmp_ind+=1
+      end
     end
 
     def to_file(filename)
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:200]: Shouldn't you write to_a? Why to_s? Is this best? Why not use compression?

File: lib/content_data/content_data.rb [hunk 27]
```diff
@@ -417,45 +262,67 @@ def validate(params = nil)
...
 
       is_valid = true
-      instances.each_value do |instance|
-        unless check_instance instance
-          is_valid = false
-
-          unless params.nil? || params.empty?
-            process_params.call :content => contents[instance.checksum], :instance => instance
+      @db.keys.each { |checksum|
+        instances = @db[checksum]
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:283]: Iterate over keys and indexes together

@db.each { |key, value|
...
}

File: lib/content_data/content_data.rb [hunk 28]
```diff
@@ -555,4 +431,314 @@ def get_query(variable, params)
 
     private :shallow_check, :deep_check, :check_instance, :get_query
   end
-end
+  # merges content data a and content data b to a new content data and returns it.
+  def self.merge(a, b)
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:435]: too much nesting, i.e., too much { { { {
To make code cleaner, do something like this:

return nil if a.nil? and b.nil?
return ContentData.new(a) if b.nil?
return ContentData.new(b) if a.nil?

# Assuming a and b are not nil....
some code here...

File: lib/content_data/content_data.rb [hunk 29]
```diff
@@ -555,4 +431,314 @@ def get_query(variable, params)
...
+        return ContentData.new(b)
+      end
+    else
+      if b.nil?
+        # A is active. B is nil.
+        return ContentData.new(a)
+      else
+        # A is active. B is active. Start merge
+        a_db = a.private_db  #db is not cloned
+        b_db = b.db  #db is cloned
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:451]: db is bad method, make a copy constructor instead.
ContentData.new(b)....

Comment (kolmanv) [inline lib/content_data/content_data.rb:451]: Consider useing private_db of a and b and iterating over them and while iterating generate new content data instance, without clone/copy of db. Creating the merged one from scratch.

File: lib/content_data/content_data.rb [hunk 30]
```diff
@@ -555,4 +431,314 @@ def get_query(variable, params)
...
+        }
+        return_cd = ContentData.new()
+        return_cd.set_db(b_db)
+        return return_cd
+      end
+    end
+  end
+
+  def self.merge_override_b(a, b)
+    if a.nil?
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:485]: Same as above. (less if and {{{{)

File: lib/content_data/content_data.rb [hunk 31]
```diff
@@ -555,4 +431,314 @@ def get_query(variable, params)
...
+  end
+
+  # returns the common content in both a and b
+  def self.intersect(a, b)
+    b_minus_a = remove(a, b)
+    b_minus_b_minus_a  = remove(b_minus_a, b)
+    return b_minus_b_minus_a
+  end
+
+  def self.unify_time(cd)
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:712]: Please describe in few lines (comment) exactly what it does.

File: lib/content_data/content_data.rb [hunk 32]
```diff
@@ -555,4 +431,314 @@ def get_query(variable, params)
...
+  end
+
+  def self.remove_directory(cd, global_dir_path)
+    return nil if cd.nil?
+    ret_cd = ContentData.new()
+    return ret_cd if cd.private_db.nil?
+    ret_db = Hash.new
+    ret_cd.set_db(ret_db)
+    return ret_cd if cd.private_db.size == 0
+    #puts "remove_directory: checking dir:#{global_dir_path}"
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:678]: Remove or add as Log.debug

File: lib/content_data/content_data.rb [hunk 33]
```diff
@@ -555,4 +431,314 @@ def get_query(variable, params)
...
+        b_db = b.private_db  #db is not cloned
+                             #loop over contents of A and add non existing contents of A in B to B
+        a_db.keys.each { |a_checksum|
+          a_instances = a_db[a_checksum]
+          b_instances = b_db[a_checksum]
+          if b_instances.nil?
+            #B does not contain current A content.
+            # Add A content to B db
+            b_db[a_checksum] = [a_instances[0],
+                                a_instances[1].clone,
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:509]: Why you need to clone this?

File: lib/content_data/content_data.rb [hunk 34]
```diff
@@ -555,4 +431,314 @@ def get_query(variable, params)
...
+                                a_instances[2]]
+          else
+            #checksum exists in both A and B.
+            #Loop over instances and add non existing instances of A in B to B
+            # TODO: do something with contents time stamps?
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:514]: TODO(owner)

File: lib/content_data/content_data.rb [hunk 35]
```diff
@@ -555,4 +431,314 @@ def get_query(variable, params)
...
+  #    Content_2 ->
+  #                   Instance_3
+  #                   Instance_4
+  #    Content_3 ->
+  #                   Instance_5
+  # B-A db:
+  #   Content_3 ->
+  #                   Instance_5
+  def self.remove(a, b)
+    if a.nil?
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:558]: Same as above..

File: lib/content_data/content_data.rb [hunk 36]
```diff
@@ -555,4 +431,314 @@ def get_query(variable, params)
...
+        b_db = b.private_db  #db is not cloned
+        c_db = Hash.new  # this is the result hash to be returned
+                             #loop over contents of B and add non existing contents of B in A to C
+        b_db.keys.each { |b_checksum|
+          a_instances = a_db[b_checksum]
+          if a_instances.nil?
+            #B content is not found in A. Add to C
+            b_instances = b_db[b_checksum]
+            c_db[b_checksum] = [b_instances[0],
+                                b_instances[1].clone,
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:583]: Why need to clone?

File: lib/content_data/content_data.rb [hunk 37]
```diff
@@ -555,4 +431,314 @@ def get_query(variable, params)
...
+  #
+  #    Content_2 ->
+  #                   Instance_3
+  #                   Instance_4
+  #    Content_3 ->
+  #                   Instance_5
+  # B-A db:
+  #   Content_3 ->
+  #                   Instance_5
+  def self.remove(a, b)
```
Comment (kolmanv) [inline lib/content_data/content_data.rb:557]: Consider adding all those methods as private methods inside ContentData class.
Consider changing ContentData to Index (lets discuss this....)

File: lib/content_server/backup_server.rb [hunk 38]
```diff
@@ -87,7 +87,7 @@ def run_backup_server
         sleep(Params['backup_check_delay'])
         local_cd = dynamic_content_data.last_content_data()
         remote_cd = content_server_dynamic_content_data.last_content_data()
-        diff = ContentData::ContentData.remove(local_cd, remote_cd)
+        diff = ContentData.remove(local_cd, remote_cd)
```
Comment (kolmanv) [inline lib/content_server/backup_server.rb:90]: If it will be class member funciton we can write
diff = remote_cd.remove(local_cd)

File: lib/content_server/queue_copy.rb [hunk 39]
```diff
@@ -50,20 +50,23 @@ def run()
...
-                @backup_tcp.send_obj([:ACK_MESSAGE, [instance.checksum, Time.now.to_i]])
-              end
-            }
-          elsif message_type == :ACK_MESSAGE
-            # Received ack from backup, copy file if all is good.
-            # The timestamp is of local content server! not backup server!
+            if message_type == :COPY_MESSAGE
+              Log.info "Copy file event: #{message_content}"
+              # Prepare source,dest map for copy.
+              message_content.private_db.keys.each {|checksum|
```
Comment (kolmanv) [inline lib/content_server/queue_copy.rb:56]: Why not |key, value| iteration?

Comment (kolmanv) [inline lib/content_server/queue_copy.rb:56]: Lets consider how we move this logic inside ContentData
Maybe we can create method called iterate which gets block and executed things on the block...

message_content.iterate { |checksum, instances|
...
}

File: lib/content_server/queue_indexer.rb [hunk 40]
```diff
@@ -17,92 +17,120 @@ def initialize(input_queue, output_queue, content_data_path)
...
-          # Add non existing and changed files to index queue.
-          @input_queue.push([FileMonitoring::FileStatEnum::STABLE, instance.full_path])
-        end
-      end
+        tmp_content_data = ContentData::ContentData.new
+        tmp_content_data.from_file(@content_data_path) if File.exists?(@content_data_path)
+        #Loop over instances from file
+        #  If file time and size matches then add it to DB
+        #  Else, monitor the path again
+        tmp_content_data.private_db.keys.each { |checksum|
```
Comment (kolmanv) [inline lib/content_server/queue_indexer.rb:25]: Same as above, hide the content data logic,
Leave only one method call + block.

File: lib/file_utils/file_utils.rb [hunk 41]
```diff
@@ -198,67 +198,85 @@ def self.contet_data_command(command, cd_a, cd_b, dest_path)
...
-            Log.info ("Comparison success.")
-            File.utime File.atime(instance.full_path), instance.modification_time, instance.full_path
-            file_mtime = File.open(instance.full_path) { |f| f.mtime }
-            Log.info "file mtime:#{file_mtime}."
-            Log.info "file mtime:#{file_mtime.to_i}."
-          end
-        end
+      #             then the assumption is that this file wasnt indexized and it will not be treated
+      #             (e.i. we do nothing with it)
+      def self.unify_time(cd)
```
Comment (kolmanv) [inline lib/file_utils/file_utils.rb:203]: We already have that implemented in content_data.rb right?

File: test/content_data/content_data_test.rb [hunk 42]
```diff
@@ -3,143 +3,353 @@
...
-      #                               "/home/kuku/dev/lala/k.txt", ContentData.parse_time("1296527039"))
-      #instance_serializer = instance.serialize()
-      #instance_copy = ContentInstance.new(nil, nil, nil, nil, nil, nil, instance_serializer)
-      #assert_equal(instance, instance_copy)
-
-      # ContentData serialization test
-      #content_data_serializer = content_data.serialize
-      #content_data_copy = ContentData.new(content_data_serializer)
-      #assert_equal(content_data, content_data_copy)
+      def test_cloning_db_1
```
Comment (kolmanv) [inline test/content_data/content_data_test.rb:7]: General comment: No files should be generated in the test!
Only memory files StringIO.
We should consider changing the interface such that it will be easy to do that.

File: test/file_utils/fileutil_mksymlink_test.rb [hunk 43]
```diff
@@ -63,32 +63,32 @@ def setup
...
-        @base_db.instances.each_value do |instance|
-          unless only_in_base.include?(instance.full_path)
-            file_name = File.basename(instance.full_path)
-            base_dir = File.dirname(instance.full_path)
-            ref_full_path = String.new(REF_DIR)
-            if (base_dir == RESOURCES_DIR)
-              ref_full_path << "/dir/#{file_name}"
-            else
-              ref_full_path << "/#{file_name}"
+        @base_db.private_db.keys.each { |checksum|
```
Comment (kolmanv) [inline test/file_utils/fileutil_mksymlink_test.rb:66]: General Note: No files should be generated in the real system.

File: test/file_utils/time_modification_test.rb [hunk 44]
```diff
@@ -9,14 +9,20 @@
...
+  def FileUtils.parse_time time_str
+    return nil unless time_str.instance_of? String
+    seconds_from_epoch = Integer time_str  # Not using to_i here because it does not check string is integer.
+    time = Time.at seconds_from_epoch
+  end
+
   module Test
     class TestTimeModification < ::Test::Unit::TestCase
       # directory where tested files will be placed: __FILE__/time_modification_test
       RESOURCES_DIR = File.expand_path(File.dirname(__FILE__) + "/time_modification_test")
```
Comment (kolmanv) [inline test/file_utils/time_modification_test.rb:21]: No files should be generated in the real file system in tests.

File: lib/content_data/content_data.rb (hunk 3 above)
Comment (genadyp) [inline lib/content_data/content_data.rb:57]: I use it in specs and remote index validations for example.
But nevertheless this method is insecure.
Yaron, could you add instead methods to go through ContentData?
Or I missed something and there is a way to do it?

File: lib/content_data/content_data.rb (hunk 7 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:127]: method: .hash - Will compute a hash-code for this hash. Two hashes with the same content
will have the same hash code (and will compare using <code>eql?</code>).

I think I need to use eq? to be sure it compares the way described above..

File: lib/content_data/content_data.rb (hunk 6 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:124]: method: == should compare all data base (this is how it was used before..)
We could think on more specific cabability such as comparing contents\instances

File: lib/content_data/content_data.rb (hunk 4 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:109]: ok. will check it with conciderations of time\memory.
We saw that memory consumption is an issue with 1M data base items.

File: lib/content_data/content_data.rb (hunk 2 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:43]: Ok

File: lib/content_data/content_data.rb (hunk 1 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:28]: Ok

File: lib/content_data/content_data.rb (hunk 8 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:130]: This is legacy code (and also good for testings and user frindly logging). 
The binary capability is not completed yet. When done, we could think of dropping the to_s

File: lib/content_data/content_data.rb (hunk 9 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:73]: already discussed above. We need to think of CPU\Memory conciderations.

File: lib/content_data/content_data.rb (hunk 12 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:44]: cool!

File: lib/content_data/content_data.rb (hunk 13 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:61]: good point
this will part of refactoring the interface, defining cloning, attributes etc

File: lib/content_data/content_data.rb (hunk 14 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:110]: finally getting to know ruby's slang! Grisha! Where have yu been till now?

File: lib/file_utils/file_utils.rb (hunk 23 above)
Comment (yarondbb) [inline lib/file_utils/file_utils.rb:278]: you mean, to remove the return and leave not_found_cd right?

File: lib/content_data/content_data.rb (hunk 24 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:83]: Ok. The TODO is for me :)

File: lib/content_data/content_data.rb (hunk 27 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:283]: From banchmark I perfromed, best time performence achived when iterating on keys and getting the value for each key. I believe the reason is that, 'each' creates 2 arrays and then iterate thorugh them, while on the other hand, each_key or each_value creat only 1 array in he background

File: lib/content_data/content_data.rb (hunk 33 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:509]: a_instances[1] is a hash object (of instances) in A data base
Without the clone (nativ shalow clone of Hash object which is all that needed in instances hash), B
will have a referece to A instances hash. This could lead to a unwanted situation when changin A will affect B..

File: lib/content_data/content_data.rb (hunk 34 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:514]: Again.. I am the owner. The question here is what do to with content time? we need to think of content data in a much wide perspective and decide.

File: lib/content_data/content_data.rb (hunk 35 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:558]: Pls elaborate? (I do not see TODO here as the above comment)

File: lib/content_data/content_data.rb (hunk 36 above)
Comment (yarondbb) [inline lib/content_data/content_data.rb:583]: Without the clone (nativ shalow clone of Hash object which is all that needed in instances hash), C
will have a referece to B instances hash. This could lead to a unwanted situation when changin B will affect C..

File: lib/content_server/queue_indexer.rb (hunk 40 above)
Comment (yarondbb) [inline lib/content_server/queue_indexer.rb:25]: I am thinking of iterators such as:

tmp_content_data.each_content { |content_info|
  checksum = content_info[:checksum]
  size = content_info[:size]
  content_time_mod = content_info[:content_time_mod]
  ..
}

and-

tmp_content_data.each_instance { |instance_info|
  checksum = instance_info[:checksum]
  size = instance_info[:size]
  instance_time_mod = instance_info[:instance_time_mod]
  server = instance_info[:server]
  device = instance_info[:device]
  file_path = instance_info[:file_path]
  ..
}

File: lib/file_utils/file_utils.rb (hunk 41 above)
Comment (yarondbb) [inline lib/file_utils/file_utils.rb:203]: No. Here we call the implementation of unify_time in content_data.rb (at line #204), then
We phisically update the time if needed

File: lib/content_server/queue_indexer.rb (hunk 40 above)
Comment (genadyp) [inline lib/content_server/queue_indexer.rb:25]: We discussed last week that iterators return Hash to block (e.g. content_info, instance_info),
but may be it will be better to work just with variables 
(e.g tmp_content_data.each_instance { |checksum, size, ...|)?
At least examples that I saw do so.

File: lib/file_indexing/index_agent.rb (hunk 19 above)
Comment (yarondbb) [inline lib/file_indexing/index_agent.rb:74]: When I removed it then the 'unless' in the next line was not related to this line..
